src/validation.cpp

```diff
@@ -1740,10 +1740,16 @@ static bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockInd
         nExpectedMint += nFees;
 
     //Check that the block does not overmint
-    if (!IsBlockValueValid(pindex->nHeight, nExpectedMint, nMint)) {
-        return state.DoS(100, error("ConnectBlock() : reward pays too much (actual=%s vs limit=%s)",
-                                    FormatMoney(nMint), FormatMoney(nExpectedMint)),
-                         REJECT_INVALID, "bad-cb-amount");
+    CAmount nBudgetAmt = 0;     // If this is a superblock, amount to be paid to the winning proposal, otherwise 0
+    if (!IsBlockValueValid(pindex->nHeight, nExpectedMint, nMint, nBudgetAmt)) {
+        return state.DoS(100, error("%s: reward pays too much (actual=%s vs limit=%s)",
+                                    __func__, FormatMoney(nMint), FormatMoney(nExpectedMint)),
+                         REJECT_INVALID, "bad-blk-amount");
+    }
+
+    // For blocks v10+: Check that the coinbase pays the exact amount
+    if (isPoSActive && pindex->nVersion >= 10 && !IsCoinbaseValueValid(block.vtx[0], nBudgetAmt)) {
+        return state.DoS(100, false, REJECT_INVALID, "bad-cb-amount");
```

Comment:
What we can do, I think, is add a validation-state distinction between regular-blocks checks, and superblocks checks. 

Since we are going to pay budgets and masternodes together during superblocks, there won't be any difference between bad-cb-amount or bad-budget amount at this level.
This particular function isn't meant to check if the masternode/payments are good. That's the job of `CBudgetManager/CMasternodePayments::IsTransactionValid`.
Here we are only looking at the total output of the coinbase tx, and checking that it is exactly as expected, no more, no less (merely to prevent abuses with cold staking, where e.g. the staker could underpay the owner, overpaying a masternode, without resulting in block "overmint").

I've refactored this bit, passing the active state to `IsCoinbaseValueValid`.
Also, I've added a brand new unit test to cover it.